File: nelsie/python/nelsie/layoutexpr.py
from typing import Union, Sequence

from .utils import check_is_type

IntOrFloatOrLayoutExpr = Union[int, float, "LayoutExpr"]


class LayoutExpr:
    def __init__(self, op, arg0=None, arg1=None, arg2=None):
        self._op = op
        self._arg0 = arg0
        self._arg1 = arg1
        self._arg2 = arg2

    def __add__(self, other: IntOrFloatOrLayoutExpr):
        check_int_or_float_or_layout_expr(other)
        return LayoutExpr("+", self, other)

    def __mul__(self, other: IntOrFloatOrLayoutExpr):
        check_int_or_float_or_layout_expr(other)
        return LayoutExpr("*", self, other)

    def __sub__(self, other: IntOrFloatOrLayoutExpr):
        check_int_or_float_or_layout_expr(other)
        return LayoutExpr("-", self, other)

    @staticmethod
    def max(expressions: Sequence["LayoutExpr"]):
        return LayoutExpr("max", expressions)

    @staticmethod
    def x(node_id):
        return LayoutExpr("x", node_id)

    @staticmethod
    def y(node_id):
        return LayoutExpr("y", node_id)

    @staticmethod
    def width(node_id, fraction):
        return LayoutExpr("width", node_id, fraction)

    @staticmethod
    def height(node_id, fraction):
        return LayoutExpr("height", node_id, fraction)

    @staticmethod
    def line_x(node_id, line_idx):
        return LayoutExpr("line_x", node_id, line_idx)

    @staticmethod
    def line_y(node_id, line_idx):
        return LayoutExpr("line_y", node_id, line_idx)

    @staticmethod
    def line_width(node_id, line_idx, fraction):
        return LayoutExpr("line_width", node_id, line_idx, fraction)

    @staticmethod
    def line_height(node_id, line_idx, fraction):
        return LayoutExpr("line_height", node_id, line_idx, fraction)

    @staticmethod
    def inline_x(node_id, anchor_id):
        return LayoutExpr("inline_x", node_id, anchor_id)

    @staticmethod
    def inline_y(node_id, anchor_id):
        return LayoutExpr("inline_y", node_id, anchor_id)

    @staticmethod
    def inline_width(node_id, anchor_id, fraction):
        return LayoutExpr("inline_width", node_id, anchor_id, fraction)

    @staticmethod
    def inline_height(node_id, anchor_id, fraction):
        return LayoutExpr("inline_height", node_id, anchor_id, fraction)


int_or_float_or_layout_expr = (int, float, LayoutExpr)


def check_int_or_float_or_layout_expr(obj):
    check_is_type(obj, int_or_float_or_layout_expr)

File: nelsie/src/pyinterface/layoutexpr.rs
use pyo3::exceptions::PyValueError;
use pyo3::types::PyAnyMethods;
use pyo3::{intern, Bound, PyAny, PyResult};
use renderer::{InlineId, LayoutExpr, NodeId};

// #[derive(Debug)]
// pub(crate) struct PyLayoutExpr(LayoutExpr);
//
// impl From<PyLayoutExpr> for LayoutExpr {
//     fn from(value: PyLayoutExpr) -> Self {
//         value.0
//     }
// }
//
// impl<'py> FromPyObject<'py> for PyLayoutExpr {
//     fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
//         Ok(crate::pyinterface::extract::PyLengthOrExpr(if let Ok(value) = obj.extract::<f32>() {
//             LengthOrExpr::points(value)
//         } else if let Ok(value) = obj.extract::<&str>() {
//         extract_layout_expr(ob.getattr("_expr")?).map(PyLayoutExpr)
//     }
// }

pub(crate) fn extract_layout_expr(obj: &Bound<PyAny>) -> PyResult<LayoutExpr> {
    if let Ok(value) = obj.extract() {
        return Ok(LayoutExpr::const_value(value));
    }
    let py = obj.py();
    let arg0 = intern!(py, "_arg0");
    let v0 = obj.getattr(arg0)?;
    let arg1 = intern!(py, "_arg1");
    let v1 = obj.getattr(arg1)?;
    let arg2 = intern!(py, "_arg2");
    let op = obj.getattr(intern!(py, "_op"))?;
    let name: &str = op.extract()?;
    match name {
        "+" | "-" | "*" => {
            let expr_a = extract_layout_expr(&v0)?;
            let expr_b = extract_layout_expr(&v1)?;
            Ok(match name {
                "+" => LayoutExpr::add(expr_a, expr_b),
                "-" => LayoutExpr::sub(expr_a, expr_b),
                "*" => LayoutExpr::mul(expr_a, expr_b),
                _ => unreachable!(),
            })
        }
        "x" => Ok(LayoutExpr::X {
            node_id: NodeId::new(v0.extract()?),
        }),
        "y" => Ok(LayoutExpr::Y {
            node_id: NodeId::new(v0.extract()?),
        }),
        "width" => Ok(LayoutExpr::Width {
            node_id: NodeId::new(v0.extract()?),
            fraction: v1.extract()?,
        }),
        "height" => Ok(LayoutExpr::Height {
            node_id: NodeId::new(v0.extract()?),
            fraction: v1.extract()?,
        }),
        "line_x" => Ok(LayoutExpr::LineX {
            node_id: NodeId::new(v0.extract()?),
            line_idx: v1.extract()?,
        }),
        "line_y" => Ok(LayoutExpr::LineY {
            node_id: NodeId::new(v0.extract()?),
            line_idx: v1.extract()?,
        }),
        "line_width" => Ok(LayoutExpr::LineWidth {
            node_id: NodeId::new(v0.extract()?),
            line_idx: v1.extract()?,
            fraction: obj.getattr(arg2)?.extract()?,
        }),
        "line_height" => Ok(LayoutExpr::LineHeight {
            node_id: NodeId::new(v0.extract()?),
            line_idx: v1.extract()?,
            fraction: obj.getattr(arg2)?.extract()?,
        }),
        "inline_x" => Ok(LayoutExpr::InlineX {
            node_id: NodeId::new(v0.extract()?),
            inline_id: InlineId::new(v1.extract()?),
        }),
        "inline_y" => Ok(LayoutExpr::InlineY {
            node_id: NodeId::new(v0.extract()?),
            inline_id: InlineId::new(v1.extract()?),
        }),
        "inline_width" => Ok(LayoutExpr::InlineWidth {
            node_id: NodeId::new(v0.extract()?),
            inline_id: InlineId::new(v1.extract()?),
            fraction: obj.getattr(arg2)?.extract()?,
        }),
        "inline_height" => Ok(LayoutExpr::InlineHeight {
            node_id: NodeId::new(v0.extract()?),
            inline_id: InlineId::new(v1.extract()?),
            fraction: obj.getattr(arg2)?.extract()?,
        }),
        "max" => Ok(LayoutExpr::max({
            let v: Vec<Bound<PyAny>> = v0.extract()?;
            v.into_iter()
                .map(|obj| extract_layout_expr(&obj))
                .collect::<PyResult<_>>()?
        })),
        _ => Err(PyValueError::new_err("Invalid expression")),
    }
}
